RecoLocalTracker/SiPixelRecHits/plugins/PixelCPEFastESProducer.cc

```diff
@@ -0,0 +1,102 @@
+#include "FWCore/Framework/interface/ModuleFactory.h"
+#include "FWCore/Framework/interface/ESProducer.h"
+
+// new record
+#include "CondFormats/DataRecord/interface/SiPixelGenErrorDBObjectRcd.h"
+
+#include <string>
+#include <memory>
+
+class PixelCPEFastESProducer : public edm::ESProducer {
+public:
+  PixelCPEFastESProducer(const edm::ParameterSet& p);
+  std::unique_ptr<PixelClusterParameterEstimator> produce(const TkPixelCPERecord&);
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  edm::ESGetToken<MagneticField, IdealMagneticFieldRecord> magfieldToken_;
+  edm::ESGetToken<TrackerGeometry, TrackerDigiGeometryRecord> pDDToken_;
+  edm::ESGetToken<TrackerTopology, TrackerTopologyRcd> hTTToken_;
+  edm::ESGetToken<SiPixelLorentzAngle, SiPixelLorentzAngleRcd> lorentzAngleToken_;
+  edm::ESGetToken<SiPixelLorentzAngle, SiPixelLorentzAngleRcd> lorentzAngleWidthToken_;
+  edm::ESGetToken<SiPixelGenErrorDBObject, SiPixelGenErrorDBObjectRcd> genErrorDBObjectToken_;
+
+  edm::ParameterSet pset_;
+  bool UseErrorsFromTemplates_;
+};
+
+using namespace edm;
+
+PixelCPEFastESProducer::PixelCPEFastESProducer(const edm::ParameterSet& p) {
+  std::string myname = p.getParameter<std::string>("ComponentName");
+  auto magname = p.getParameter<edm::ESInputTag>("MagneticFieldRecord");
+  UseErrorsFromTemplates_ = p.getParameter<bool>("UseErrorsFromTemplates");
+
+  pset_ = p;
+  auto c = setWhatProduced(this, myname);
+  c.setConsumes(magfieldToken_, magname)
+      .setConsumes(pDDToken_)
+      .setConsumes(hTTToken_)
+      .setConsumes(lorentzAngleToken_, edm::ESInputTag(""));
+  c.setConsumes(lorentzAngleWidthToken_, edm::ESInputTag("", "forWidth"));
+  if (UseErrorsFromTemplates_) {
+    c.setConsumes(genErrorDBObjectToken_);
+  }
+}
+
+std::unique_ptr<PixelClusterParameterEstimator> PixelCPEFastESProducer::produce(const TkPixelCPERecord& iRecord) {
+  // add the new la width object
+  const SiPixelLorentzAngle* lorentzAngleWidthProduct = nullptr;
+  lorentzAngleWidthProduct = &iRecord.get(lorentzAngleWidthToken_);
+
+  const SiPixelGenErrorDBObject* genErrorDBObjectProduct = nullptr;
+
+  // Errors take only from new GenError
+  if (UseErrorsFromTemplates_) {  // do only when generrors are needed
+    genErrorDBObjectProduct = &iRecord.get(genErrorDBObjectToken_);
+    //} else {
+    //std::cout<<" pass an empty GenError pointer"<<std::endl;
+  }
+  return std::make_unique<PixelCPEFast>(pset_,
+                                        &iRecord.get(magfieldToken_),
+                                        iRecord.get(pDDToken_),
+                                        iRecord.get(hTTToken_),
+                                        &iRecord.get(lorentzAngleToken_),
+                                        genErrorDBObjectProduct,
+                                        lorentzAngleWidthProduct);
+}
+
+void PixelCPEFastESProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
```

Comment:
The existing CPE ESProducers use inheritance to define the configuration in steps, but I am not expert enough in how the original code is structured, so I implemented a stand-alone `fillDescriptions()` method.

To be reviewed by Tracker DPG along with the rest ...

Reply:
This class looks a copy paste of https://github.com/cms-sw/cmssw/blob/master/RecoLocalTracker/SiPixelRecHits/plugins/PixelCPEGenericESProducer.cc.
You can get rid of some of these using a call to ` PixelCPEBase::fillPSetDescription(desc);`.
For the algorithm-specific ones, we implemented https://github.com/cms-sw/cmssw/blob/91be707d4e28bfd3b85525a5feae0e36b0f39181/RecoLocalTracker/SiPixelRecHits/src/PixelCPEGeneric.cc#L748
I believe the same should be applied to the `PixelCPEFast` class.